rust/src/proofs/types.rs

```diff
@@ -169,6 +169,44 @@ impl Drop for fil_PoStProof {
     }
 }
 
+#[repr(C)]
+pub struct fil_VanillaProof {
+    pub proof_len: libc::size_t,
+    pub proof_ptr: *const u8,
+}
+
+impl Clone for fil_VanillaProof {
+    fn clone(&self) -> Self {
+        let mut clone = fil_VanillaProof {
+            proof_len: 0,
+            proof_ptr: ptr::null(),
+        };
+
+        unsafe {
+            clone.proof_len = self.proof_len;
+            clone.proof_ptr = libc::malloc(self.proof_len) as *const u8;
+
+            libc::memcpy(
+                clone.proof_ptr as *mut core::ffi::c_void,
+                self.proof_ptr as *const core::ffi::c_void,
+                self.proof_len,
+            );
```

Comment:
while the above works, it should be safer and less error prone to do

```rust
let slice: &[u8] = unsafe { std::slice::from_raw_parts(self.proof_ptr, self.proof_len) };
let cloned: Box<[u8]> = slice.to_vec().into_boxed_slice();
debug_assert_eq!(self.proof_len, cloned.len());

fil_VanillaProof {
  proof_len: self.proof_len,
  proof_ptr: Box::into_raw(cloned),
}
```

this way it is guranteed that reconstructing this will actually work as expected